Fix performance bug in buildLocationList

tmsri opened the pull request:
In buildLocationList, with basic block sections, we iterate over
every basic block twice to detect section start and end. This is
sub-optimal and shows up as significantly time consuming when
compiling large functions.

This patch uses the set of sections already stored in MBBSectionRanges
and iterates over sections rather than basic blocks.

When detecting if loclists can be merged, the end label of an entry is
matched with the beginning label of the next entry. For the section
corresponding to the entry basic block, this is skipped. This is
because the loc list uses the end label corresponding to the function
whereas the MBBSectionRanges map uses the function end label.

For example:

.Lfunc_begin0:
.file
.loc 0 4 0 # ex2.cc:4:0
.cfi_startproc
.Ltmp0:
.loc 0 8 5 prologue_end # ex2.cc:8:5
....
.LBB_END0_0:
.cfi_endproc
.section .text._Z4testv,"ax",@progbits,unique,1
...
.Lfunc_end0:
.size _Z4testv, .Lfunc_end0-_Z4testv

The debug loc uses ".LBB_END0_0" for the end of the section whereas
MBBSectionRanges uses ".Lfunc_end0".

It is alright to skip this as we already check the section corresponding
to the debugloc entry.

Added a new test case to check that if this works correctly when the
variable's value is mutated in the entry section.

tmsri committed:
Fix performance bug in buildLocationList

In buildLocationList, with basic block sections, we iterate over
every basic block twice to detect section start and end. This is
sub-optimal and shows up as significantly time consuming when
compiling large functions.

This patch uses the set of sections already stored in MBBSectionRanges
and iterates over sections rather than basic blocks.

When detecting if loclists can be merged, the end label of an entry is
matched with the beginning label of the next entry. For the section
corresponding to the entry basic block, this is skipped. This is
because the loc list uses the end label corresponding to the function
whereas the MBBSectionRanges map uses the function end label.

For example:

.Lfunc_begin0:
.file
.loc 0 4 0 # ex2.cc:4:0
.cfi_startproc
.Ltmp0:
.loc 0 8 5 prologue_end # ex2.cc:8:5
....
.LBB_END0_0:
.cfi_endproc
.section .text._Z4testv,"ax",@progbits,unique,1
...
.Lfunc_end0:
.size _Z4testv, .Lfunc_end0-_Z4testv

The debug loc uses ".LBB_END0_0" for the end of the section whereas
MBBSectionRanges uses ".Lfunc_end0".

It is alright to skip this as we already check the section corresponding
to the debugloc entry.

Added a new test case to check that if this works correctly when the
variable's value is mutated in the entry section.

tmsri commented:
Continuation from PR [108886](https://github.com/llvm/llvm-project/pull/108886) I added a test case to address jmorse's recent comment.

tmsri committed:
Const value variable within a scope test.

Test to check that a variable declared within a scope that has basic block
sections still produces DW_AT_const_value.

tmsri commented:
I added one more test where the variable is only valid within a lexical scope that is part of the main section and another basic block section.

tmsri commented:
Ping.  I believe one concern with this patch was about a variable that is only valid within a lexical scope which is defined in the entry block's section.  I added a test to cover that case.   Is there any other concern?

amharc commented:
I think the logic in this PR is correct: the comment for `buildLocationList` says "If the resulting list has only one entry that is valid throughout variable's scope return true" and this seems to be the case with this PR too. I'm not a debug info expert though.

jmorse reviewed (commented):
Eyeballing this further, I was under the impression that this code was evaluating the validity of issuing a single-location in general, but I see now that's achieved by the earlier call to `validThroughout`. I see now that this is more about verifying the basic-block-section ranges are suitable to merge once the variable is identified as having a single-location, such as a constant value, is that correct?

I think my difficulty is in seeing any situation where the code from line 8130 onwards should return false -- if `validThroughout` has returned true, then we can use a single-location that implicitly covers every source-location that's in scope without the need for a location list surely? (I imagine it's tricky to identify the instruction-ranges that are in scope with basic block sections, but that's not a matter for buildLocationList).

Stepping back even further, I'm sort of questioning why this code is here without having a real handle on it; for this PR itself the change looks correct, with one inline question.

jmorse commented:
This continued to bug me a bit, so I put some traps on the `return false` paths at the end of `buildLocationList` with this PR and didn't get any test failures when running `ninja check-llvm`. Sorry to ask for more, but a test that covers that path would really help my understanding.

https://github.com/jmorse/llvm-project/commit/7b78ea87febc00b77ce8000659f61288c531eb09 <--- based on 
11a5026 with the addition of some x86 traps, which should cause tests covering those paths to fail.

tmsri commented:
> Eyeballing this further, I was under the impression that this code was evaluating the validity of issuing a single-location in general, but I see now that's achieved by the earlier call to `validThroughout`. I see now that this is more about verifying the basic-block-section ranges are suitable to merge once the variable is identified as having a single-location, such as a constant value, is that correct?
> 

Yes, this is correct.  Even if it is not merged, it will produce something that is valid.

tmsri commented on llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp:
```diff
@@ -1824,22 +1821,27 @@ bool DwarfDebug::buildLocationList(SmallVectorImpl<DebugLocEntry> &DebugLoc,
-    // merged.
-    if (CurEntry->getEndSym() != RangeMBB->getEndSymbol() ||
-        NextEntry->getBeginSym() != RangeMBB->getNextNode()->getSymbol() ||
+    // merged.  Do not match the section label end if it is the entry block
+    // section.  This is because the end label for the Debug Loc and the
+    // Function end label could be different.
+    if ((RangeIt->second.EndLabel != Asm->getFunctionEnd() &&
+         CurEntry->getEndSym() != RangeIt->second.EndLabel) ||
+        NextEntry->getBeginSym() != NextRangeIt->second.BeginLabel ||
```
Basically, MBBSectionRanges stores the ranges for each section.  For the entry section, the end label is the Function End Label, Asm->getFunctionEnd().  The entry section is actually the original function section.

Now, take a look at this assembly:

```
	.section	.text._Z4testv,"ax",@progbits
	.globl	_Z4testv                        # -- Begin function _Z4testv
	.type	_Z4testv,@function
_Z4testv:                               # @_Z4testv
.Lfunc_begin0:
	.loc	0 3 0                           # ex4.cc:3:0
# %bb.0:                                # %entry
	pushq	%rax
.Ltmp0:
	#DEBUG_VALUE: test:i <- 7
	.loc	0 7 5 prologue_end              # ex4.cc:7:5
	callq	_Z2f1v@PLT
        ...
.LBB_END0_0:
	.cfi_endproc
	.section	.text._Z4testv,"ax",@progbits,unique,1
_Z4testv.__part.1:                      # %if.then
	.section	.text._Z4testv,"ax",@progbits
.Lfunc_end0:
```

The entry block section uses label ".Lfunc_end0" which is Asm->getFunctionEnd(). There is an additional label that is created immediately after the end of the entry section ".LBB_END0_0" which is what the debug loc uses, like an alias.  In this case and this case alone, these two labels won't match even though semantically, they are the same.  So, I exclude that match.  

CurEntry and entry block are different things.  Sorry for the use of the overloaded entry.

jmorse replied:
Cool -- awkward terminology, but I can see what's going on.

tmsri commented:
> This continued to bug me a bit, so I put some traps on the `return false` paths at the end of `buildLocationList` with this PR and didn't get any test failures when running `ninja check-llvm`. Sorry to ask for more, but a test that covers that path would really help my understanding.
> 
> [jmorse@7b78ea8](https://github.com/jmorse/llvm-project/commit/7b78ea87febc00b77ce8000659f61288c531eb09) <--- based on [11a5026](https://github.com/llvm/llvm-project/commit/11a50269e82b6dce49249c5cbe3a989b06f0848f) with the addition of some x86 traps, which should cause tests covering those paths to fail.

Ah nice!, let me see if I can make a test for this.

tmsri committed:
Two more test cases where DW_AT_const cannot be used.

These test cases check loclist merging in buildLocationList with
basic block sections. 8.ll does not create any basic block sections
and the merging must quit early because there is no constant value.
9.11 creates sections but cannot be merged as i is not a constant value.

tmsri commented:
I added two more test cases which check the traps you wanted to test for the code snippet below.  The first one will check the first return as there aren't any basic block sections.  The second test will check the second "return false". 

```C++
while (NextEntry != DebugLoc.end()) {
    if (NextRangeIt == Asm->MBBSectionRanges.end())
      return false;
    // CurEntry should end the current section and NextEntry should start
    // the next section and the Values must match for these two ranges to be
    // merged.  Do not match the section label end if it is the entry block
    // section.  This is because the end label for the Debug Loc and the
    // Function end label could be different.
    if ((RangeIt->second.EndLabel != Asm->getFunctionEnd() &&
         CurEntry->getEndSym() != RangeIt->second.EndLabel) ||
        NextEntry->getBeginSym() != NextRangeIt->second.BeginLabel ||
        CurEntry->getValues() != NextEntry->getValues())
      return false;
```

jmorse reviewed (approved):
LGTM -- many thanks for the extra test coverage, I think that clears up my remaining questions about what's going on.
